### ISSUE TITLE
an additional LDX#$00 is added in dli_routine without any practical reason?

### ISSUE THREAD

**Issue by yaronnir2006** [cc65/cc65#910]

**Comment by oliverschmidt** [comment]
Please elaborate what "dli_routine" refers to. I can't find any reference to that term in cc65.

**Comment by yaronnir2006** [comment]
appologizes, i was in the middle of typing this.....

**Comment by yaronnir2006** [comment]
i have a dli_routine:
void dli_routine(void)
{
	asm("pha");
	ANTIC.wsync = 33;
	ANTIC.wsync = 33;
	ANTIC.wsync = 33;
	GTIA_WRITE.colbk= ANTIC.wsync = OS.color4;
	asm("pla");
	asm("rti");
}

when i look at the generated asm after compiling the ld65:
; ---------------------------------------------------------------
; void __near__ dli_routine (void)
; ---------------------------------------------------------------

.segment	"CODE"

.proc	_dli_routine: near

.segment	"CODE"

	pha
	lda     #$21
	sta     $D40A
	sta     $D40A
	sta     $D40A
	ldx     #$00
	lda     $02C8
	sta     $D40A
	sta     $D01A
	pla
	rti

.endproc


notice there is an LDX#$00 there, which i don't understand its purpose

it by the way , causes strange behavior. if i removed it everything works as expected.

any ideas?

**Comment by mrdudz** [comment]
looks like a sideeffect of integer promotion to me - is this with optimizations enabled?

(oh yeah, and you cant use C in interrupt handlers like this. problems are expected)

**Comment by yaronnir2006** [comment]
As i use debug information i've removed the optimization option.
The compiler is not aware this is an interrupt handler. it was just testing something.

edit: i tried with optimization and same LDX appears.
what does "a sideeffect of integer promotion" mean?

**Comment by yaronnir2006** [comment]
so, if i use an inline asm to push and pop the X register as well, then the generated asm does not include the LDX#$00
i am still convinced there is an issue here....


void dli_routine(void)
{
	asm("pha");
	asm("txa");
	asm("pha");
	ANTIC.wsync = 33;
	ANTIC.wsync = 33;
	ANTIC.wsync = 33;
	GTIA_WRITE.colbk = ANTIC.wsync;
	GTIA_WRITE.colbk= ANTIC.wsync = OS.color4;
	asm("pla");
	asm("tax");
	asm("pla");
	asm("rti");
}

**Comment by yaronnir2006** [comment]
any1 has an idea? can you confirm this is an issue/non issue?

**Comment by oliverschmidt** [comment]
> oh yeah, and you cant use C in interrupt handlers like this. problems are expected

Maybe I'm missing a point but as long as there are no local (aka automatic) variables defined and no functions (with more than one parameter and/or "complicated return type) called I don't see this conceptually being broken. The issue is that the compiler tends to generate calls to run time functions without the user writing a function call into the source code. Therefore the generated code has to be checked for JSR instructions.

**Comment by yaronnir2006** [comment]
well, this dli routine is not called explicitly, instead it is being placed in the vdslist location as follows:
void init_dli(void)
{
      OS.vdslst = &dli_routine;
}

the jsr to that routine can only be examined in altirra debugger. so i am not really sure how can this be checked
i am still puzzeld on why the ldx#01 was added.

**Comment by oliverschmidt** [comment]
The source

```
#include <atari.h>

void dli_routine(void)
{
  asm("pha");
  ANTIC.wsync = 33;
  ANTIC.wsync = 33;
  ANTIC.wsync = 33;
  GTIA_WRITE.colbk= ANTIC.wsync = OS.color4;
  asm("pla");
  asm("rti");
}
```

comiled using the current cc65 HEAD with

```
cc65 -t atari test.c
```

results for me in

```
.proc	_dli_routine: near

.segment	"CODE"

	pha
	ldx     #$00
	lda     #$21
	sta     $D40A
	ldx     #$00
	lda     #$21
	sta     $D40A
	ldx     #$00
	lda     #$21
	sta     $D40A
	ldx     #$00
	lda     $02C8
	sta     $D40A
	sta     $D01A
	pla
	rti
	rts

.endproc
```

so I see the need for you to elaborate how you get to assembly code with exactly one `ldx #$00`.

Apart from that you see that in general `ldx #$00` isn't something weird but rather to be expected - as @mrdudz alreaady hinted. Combine https://github.com/cc65/wiki/wiki/The-primary-register with e.g. https://stackoverflow.com/questions/46073295/implicit-type-promotion-rules to understand the background.

**Comment by oliverschmidt** [comment]
> the jsr to that routine can only be examined in altirra debugger. so i am not really sure how can this be checked

You totally misunderstood me. I wasn't talking about a 'JSR' _TO_ dli_routine but a 'JSR' _INSIDE_ dli_routine.

**Comment by groessler** [comment]
> i am still puzzeld on why the ldx#01 was added.
I'm not considering that as an "issue" (iow "bug"), but just some suboptimal code generation.

The proper way to use C functions in interrupt handlers is to have assembly language interrupt enter and exit routines which save and restore the registers (which in case of cc65 also include the ZP vars used by the compiler). And you need to be careful not to call functions which aren't reentrancy safe.

**Comment by oliverschmidt** [comment]
> i tried with optimization and same LDX appears.

I'd say this seem to be not exactly correct because _ONLY_ with optimization you get from 4 `ldx #$00` down to 1 `ldx #$00` so the assembly code you posted initially is _EXACTLY_ with optimizations.

I guess by now you see yourself that the issue at hand is that the optimizer works imperfectly. Not more, not less.

**Comment by yaronnir2006** [comment]
hmmm so perhaps this relies on the optimizer after all, where redundency happended a trail of the last one (the one remaining ldx#$00) was forgotten....perhaps that is the issue then?

**Comment by yaronnir2006** [comment]
> The proper way to use C functions in interrupt handlers is to have assembly language interrupt enter and exit routines which save and restore the registers (which in case of cc65 also include the ZP vars used by the compiler). And you need to be careful not to call functions which aren't reentrancy safe.

yes ,as i shown above when pushing a poping (using inline asm) the X register as well, it was removed from the generated asm file.

**Comment by yaronnir2006** [comment]
> > i tried with optimization and same LDX appears.
> 
> I'd say this seem to be not exactly correct because _ONLY_ with optimization you get from 4 `ldx #$00` down to 1 `ldx #$00` so the assembly code you posted initially is _EXACTLY_ with optimizations.
> 
> I guess by now you see yourself that the issue at hand is that the optimizer works imperfectly. Not more, not less.

the overall xex file created (in the main cl65 command line) doesn't have an optimizer option, BUT, i tried to individually compile the dli.c file and there i used the optimizer, so you are correct in your assumptions.

and again, this leads me to beleive there might be a bug (minor) in the optimizer....

**Comment by groessler** [comment]
> ...when pushing a poping (using inline asm)

Don't use inline asm for that. Use a read assembler source file. No unexpected side effects then.

**Comment by groessler** [comment]
> ...there might be a bug (minor) in the optimizer...

IMO this doesn't qualify as bug since the generated code is correct. It's just that the optimizer could do better in this example.

**Comment by oliverschmidt** [comment]
```
#include <atari.h>

void dli_routine(void)
{
  ANTIC.wsync = 33;
  ANTIC.wsync = 33;
  ANTIC.wsync = 33;
  GTIA_WRITE.colbk= ANTIC.wsync = OS.color4;
}
```

results in the expected

```
.proc	_dli_routine: near

.segment	"CODE"

	lda     #$21
	sta     $D40A
	sta     $D40A
	sta     $D40A
	sta     $D01A
	lda     $02C8
	sta     $D40A
	sta     $D01A
	rts

.endproc
```

so I'd say the optimizer doesn't get along that well with inline assembly.

https://cc65.github.io/doc/cc65.html#s9 says

_Note: Inline assembler statements are subject to all optimizations done by the compiler. There currently is no way to protect an inline assembler statement -- alone -- from being moved or removed completely by the optimizer. If in doubt, check the generated assembler output; or, disable optimizations (for that function)._

so I wouldn't classify your experience as a bug. Rather you need to understand that cc65 inline asm isn't really as useful as many novice cc65 users think. You can't really combine inline asm with the optimizer - and without the optimizer the C code is resulting in asm code you don't want in areas so important to you that you started to use inline asm in the first place.

In case you didn't use inline asm for optimization purposes but rather to "just" be able to write an interrupt function then you need to use https://cc65.github.io/doc/funcref.html#set_irq instead of fiddling with inline asm.

**Comment by groessler** [comment]
Nitpicking, @oliverschmidt, you seem to have a copy'n'paste error in your example. There is a spurious STA $D01A in the middle which I don't get when I compile with "cc65 -O ..."

**Comment by oliverschmidt** [comment]
Yep, I fiddled with the source - and haven't copied the right one - sorry...

**Comment by mrdudz** [comment]
> > oh yeah, and you cant use C in interrupt handlers like this. problems are expected
> 
> Maybe I'm missing a point but as long as there are no local (aka automatic) variables defined and no functions (with more than one parameter and/or "complicated return type) called I don't see this conceptually being broken. The issue is that the compiler tends to generate calls to run time functions without the user writing a function call into the source code. Therefore the generated code has to be checked for JSR instructions.

yes, i was generalizing. if you don't do this and that and have detailed knowledge about how code is generated and take the necessary precautions, then it might work. this doesnt seem to be the case here though :)

**Comment by oliverschmidt** [comment]
> this doesnt seem to be the case here though :)

:-)

**Comment by greg-king5** [comment]
> So, I'd say the optimizer doesn't get along that well with inline assembly.

The `rti` causes what we see there.  The optimizer knows what it does (it returns from interrupt handlers).  The optimizer knows that those handlers must preserve everything.  Therefore, it doesn't remove instructions that do that preservation.  It doesn't remove the instructions that put the registers on the stack, and the instructions that pull them back off the stack.  The optimizer thinks that the last `ldx #$00` is a part of the code that restores a register.  But, if the x-index register _explicitly_ is pulled from the stack, then the optimizer can know that `ldx #$00` isn't a part of the restoration.

**Comment by oliverschmidt** [comment]
@greg-king5: Thanks for the very helpful correction of our rather dull assumptions :-)

So if I got you right everything works as designed. The cc65 optimizer does exactly what can be expected by this kind of optimizer.

**Comment by greg-king5** [comment]
> So, if I got you right, everything works as designed. The cc65 optimizer does exactly what can be expected by this kind of optimizer.

Yep.

**Comment by yaronnir2006** [comment]
@greg-king5 , thank you kindly for dipicting the situation here. at first it looks as an optimizer issue, but reading your comments, i know understand why it works like that. 
so good practice would be to push and pull the  register in these type of routines if the generated code accidently "forgets" to remove an LDX/LDY instruction..
many thanks!

**Comment by greg-king5** [comment]
By the way, interrupt handling (by a C function) is a case where a fix for bug #897 would be useful.  Then, the example could be written as this:
```c
#include <atari.h>

void dli_routine(void)
{
    asm("pha");
    asm("stx %g + 1", label);
    ANTIC.wsync = 33;
    ANTIC.wsync = 33;
    ANTIC.wsync = 33;
    GTIA_WRITE.colbk = ANTIC.wsync;
    GTIA_WRITE.colbk= ANTIC.wsync = OS.color4;
label:
    asm("ldx #$00");
    asm("pla");
    asm("rti");
}
```

**Comment by groessler** [comment]
A fix for bug #897 would be nice, sure.

But, Greg, I think your suggestion is not really user friendly. How does someone know that the X register has to be preserved?

What I said before, if you really want to write interrupt handlers in C, make ASM wrappers to save and restore the machine state. Anything else is just like living on the edge.

And, honestly, I don't understand why the OP's dli_routine example should be written in C. It's simple enough to be written in assembler.

Use the tool which fits best. C for complex things, ASM for simple things. (Or ASM for complex things if you can handle it. :-) )
